scala/scala_import.bzl

```diff
@@ -1,121 +1,105 @@
-
-def _transitively_accumulate_labels(dep_target, java_provider, jars2labels):
-  if hasattr(dep_target, "jars_to_labels"):
-    jars2labels.update(dep_target.jars_to_labels)
-  #scala_library doesn't add labels to the direct dependency itself
-  for jar in java_provider.compile_jars.to_list():
-    jars2labels[jar.path] = dep_target.label
-
-def _collect_runtime(runtime_deps):
-  jar_deps = []
-  for dep_target in runtime_deps:
-      java_provider = dep_target[java_common.provider]
-      jar_deps.append(java_provider.transitive_runtime_jars)
-
-  return depset(transitive = jar_deps)
-
-def _collect_exports(exports):
-  exported_jars = depset()
-
-  for dep_target in exports:
-      java_provider = dep_target[java_common.provider]
-      exported_jars += java_provider.full_compile_jars
-
-  return exported_jars
+
+    return [
+        default_info,
+        _scala_import_java_info(ctx, direct_binary_jars),
+        _scala_import_jars_to_labels(ctx, direct_binary_jars),
+    ]
+
+def _scala_import_java_info(ctx, direct_binary_jars):
+    # merge all deps, exports, and runtime deps into single JavaInfo instances
+
+    s_deps = java_common.merge(_collect(JavaInfo, ctx.attr.deps))
+    s_exports = java_common.merge(_collect(JavaInfo, ctx.attr.exports))
+    s_runtime_deps = java_common.merge(_collect(JavaInfo, ctx.attr.runtime_deps))
+
+    # build up our final JavaInfo provider
+
+    compile_time_jars = depset(
+        direct = direct_binary_jars,
+        transitive = [
+            s_exports.transitive_compile_time_jars])
+
+    transitive_compile_time_jars = depset(
+        transitive = [
+            compile_time_jars,
+            s_deps.transitive_compile_time_jars,
+            s_exports.transitive_compile_time_jars])
+
+    transitive_runtime_jars = depset(
+        transitive = [
+            compile_time_jars,
+            s_deps.transitive_runtime_jars,
+            s_exports.transitive_runtime_jars,
+            s_runtime_deps.transitive_runtime_jars])
+
+    return java_common.create_provider(
+        ctx.actions,
+        use_ijar = False,
+        compile_time_jars = compile_time_jars,
+        transitive_compile_time_jars = transitive_compile_time_jars,
+        transitive_runtime_jars = transitive_runtime_jars)
+
+def _scala_import_jars_to_labels(ctx, direct_binary_jars):
+    # build up JarsToLabels
+    # note: consider moving this to an aspect
+
+    lookup = {}
+    for jar in direct_binary_jars:
+        lookup[jar.path] = ctx.label
+
+    for entry in ctx.attr.deps:
+        if JavaInfo in entry:
+            for jar in entry[JavaInfo].compile_jars:
+                lookup[jar.path] = entry.label
+        if JarsToLabels in entry:
+            lookup.update(entry[JarsToLabels].lookup)
+
+    for entry in ctx.attr.exports:
+        if JavaInfo in entry:
+            for jar in entry[JavaInfo].compile_jars:
+                lookup[jar.path] = entry.label
```

Comment:
if we won't change this to `ctx.label` (per exports discussion) then can we refactor these two loops to a common method? if `ctx.label` doesn't change they're exactly the same